[Quant][DBR] Add test for serialization

andrewor14 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #70078

Summary: This commit adds a serialization test for DBR.

Test Plan: python test/test_quantization.py TestQuantizeDBR.test_serialization

Reviewers: vkuzo

Subscribers: vkuzo, supriyar

Tasks: https://github.com/pytorch/pytorch/issues/69535

Differential Revision: [D33405192](https://our.internmc.facebook.com/intern/diff/D33405192)

andrewor14 committed:
[Quant][DBR] Add test for serialization

Summary: This commit adds a serialization test for DBR.

Test Plan: python test/test_quantization.py TestQuantizeDBR.test_serialization

Reviewers: vkuzo

Subscribers: vkuzo, supriyar

Tasks: https://github.com/pytorch/pytorch/issues/69535

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][DBR] Add test for serialization"


Summary: This commit adds a serialization test for DBR.

Test Plan: python test/test_quantization.py TestQuantizeDBR.test_serialization

Reviewers: vkuzo

Subscribers: vkuzo, supriyar

Tasks: https://github.com/pytorch/pytorch/issues/69535

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][DBR] Add test for serialization"


Summary: This commit adds a serialization test for DBR.

Test Plan: python test/test_quantization.py TestQuantizeDBR.test_serialization

Reviewers: vkuzo

Subscribers: vkuzo, supriyar

Tasks: https://github.com/pytorch/pytorch/issues/69535

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][DBR] Add test for serialization"


Summary: This commit adds a serialization test for DBR.

Test Plan: python test/test_quantization.py TestQuantizeDBR.test_serialization

Reviewers: vkuzo

Subscribers: vkuzo, supriyar

Tasks: https://github.com/pytorch/pytorch/issues/69535

[ghstack-poisoned]

vkuzo commented on test/quantization/dbr/test_quantize_dbr.py:
```diff
@@ -1122,6 +1123,46 @@ def test_qconfig_dict_module_name(self):
+        m = _quantize_dbr.convert(m)
+        m_copy = copy.deepcopy(m)
+
+        # Ensure the two models have different weights
```
One thing to note is that currently module swaps are handled by Eager mode quantization, and DBR reuses that code.  The "new" code that DBR adds is automatic insertion of quants and dequants, and quantization of functions and methods.  The current test mostly tests the modules.  We could add two things here to make sure we also test the DBR-only logic:
1. we can calibrate with sample data before converting, this will ensure that the scale and zero_point of observers throughout the model are based on calibration data instead of default
2. if we change things manually to make the 2nd model be different from the 1st model, we should also change the internal state of observers which will be swapped by quants.  Or, a faster option could be to create the 2nd model from scratch but do not calibrate it.

Something like this could ensure that DBR-only functionality is also being tested:

```
m = M().eval()
m = _quantize_dbr.prepare(m, qconfig_dict, example_inputs)
# calibrate, to populate statistics
m(torch.randn(1, 1, 1, 1))
m = _quantize_dbr.convert(m)

# create new model
m2 = M().eval()
m2 = _quantize_dbr.prepare(m2, qconfig_dict, example_inputs)
# do not calibrate, to ensure important statistics are populated without calibration and 
# the results are different at every node, including the quantize_per_tensor node
m2 = _quantize_dbr.convert(m2)

# now, m and m2 are different, load the state dict and verify equivalency
```

andrewor14 replied:
That makes sense. I updated the code accordingly. However, the test now fails. It seems there's something that we're not capturing in the `state_dict`, since I verified that both `state_dict`s are identical after serialization. I am investigating this further.

andrewor14 committed:
Update on "[Quant][DBR] Add test for serialization"


Summary: This commit adds a serialization test for DBR.

Test Plan: python test/test_quantization.py TestQuantizeDBR.test_serialization

Reviewers: vkuzo

Subscribers: vkuzo, supriyar

Tasks: https://github.com/pytorch/pytorch/issues/69535

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][DBR] Add test for serialization"


Summary: This commit adds a serialization test for DBR.

Test Plan: python test/test_quantization.py TestQuantizeDBR.test_serialization

Reviewers: vkuzo

Subscribers: vkuzo, supriyar

Tasks: https://github.com/pytorch/pytorch/issues/69535

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][DBR] Add test for serialization"


Summary: This commit adds a serialization test for DBR.

Test Plan: python test/test_quantization.py TestQuantizeDBR.test_serialization

Reviewers: vkuzo

Subscribers: vkuzo, supriyar

Tasks: https://github.com/pytorch/pytorch/issues/69535

[ghstack-poisoned]

vkuzo commented on torch/ao/quantization/_dbr/quantization_state.py:
```diff
@@ -117,6 +117,15 @@ def __init__(
         self.needs_dtype_transform_on_outputs = True
 
+    def get_extra_state(self):
+        return self.tensor_id_to_scale_zp
```
might be valuable to define this as a dictionary, `{'tensor_id_to_scale_zp': self.tensor_id_to_scale_zp}`.  This way if a future PR needs to add more things here, it can be done in a backward compatible way.

andrewor14 committed:
Update on "[Quant][DBR] Add test for serialization"


Summary: This commit adds a serialization test for DBR.

Test Plan: python test/test_quantization.py TestQuantizeDBR.test_serialization

Reviewers: vkuzo

Subscribers: vkuzo, supriyar

Tasks: https://github.com/pytorch/pytorch/issues/69535

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][DBR] Add test for serialization"


Summary: This commit adds a serialization test for DBR.

Test Plan: python test/test_quantization.py TestQuantizeDBR.test_serialization

Reviewers: vkuzo

Subscribers: vkuzo, supriyar

Tasks: https://github.com/pytorch/pytorch/issues/69535

[ghstack-poisoned]

vkuzo reviewed (approved):
looks great, thanks for fixing this

andrewor14 committed:
Update on "[Quant][DBR] Add test for serialization"


Summary: This commit adds a serialization test for DBR.

Test Plan: python test/test_quantization.py TestQuantizeDBR.test_serialization

Reviewers: vkuzo

Subscribers: vkuzo, supriyar

Tasks: https://github.com/pytorch/pytorch/issues/69535

Differential Revision: [D33405192](https://our.internmc.facebook.com/intern/diff/D33405192)

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][DBR] Add test for serialization"


Summary: This commit adds a serialization test for DBR.

Test Plan: python test/test_quantization.py TestQuantizeDBR.test_serialization

Reviewers: vkuzo

Subscribers: vkuzo, supriyar

Tasks: https://github.com/pytorch/pytorch/issues/69535

Differential Revision: [D33405192](https://our.internmc.facebook.com/intern/diff/D33405192)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33405192).

andrewor14 committed:
Update on "[Quant][DBR] Add test for serialization"


Summary: This commit adds a serialization test for DBR.

Test Plan: python test/test_quantization.py TestQuantizeDBR.test_serialization

Reviewers: vkuzo

Subscribers: vkuzo, supriyar

Tasks: https://github.com/pytorch/pytorch/issues/69535

Differential Revision: [D33405192](https://our.internmc.facebook.com/intern/diff/D33405192)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33405192).

andrewor14 committed:
Update on "[Quant][DBR] Add test for serialization"


Summary: This commit adds a serialization test for DBR.

Test Plan: python test/test_quantization.py TestQuantizeDBR.test_serialization

Reviewers: vkuzo

Subscribers: vkuzo, supriyar

Tasks: https://github.com/pytorch/pytorch/issues/69535

Differential Revision: [D33405192](https://our.internmc.facebook.com/intern/diff/D33405192)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33405192).

andrewor14 committed:
Update on "[Quant][DBR] Add test for serialization"


Summary: This commit adds a serialization test for DBR.

Test Plan: python test/test_quantization.py TestQuantizeDBR.test_serialization

Reviewers: vkuzo

Subscribers: vkuzo, supriyar

Tasks: https://github.com/pytorch/pytorch/issues/69535

Differential Revision: [D33405192](https://our.internmc.facebook.com/intern/diff/D33405192)

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][DBR] Add test for serialization"


Summary: This commit adds a serialization test for DBR.

Test Plan: python test/test_quantization.py TestQuantizeDBR.test_serialization

Reviewers: vkuzo

Subscribers: vkuzo, supriyar

Tasks: https://github.com/pytorch/pytorch/issues/69535

Differential Revision: [D33405192](https://our.internmc.facebook.com/intern/diff/D33405192)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33405192).

andrewor14 committed:
Update on "[Quant][DBR] Add test for serialization"


Summary: This commit adds a serialization test for DBR.

Test Plan: python test/test_quantization.py TestQuantizeDBR.test_serialization

Reviewers: vkuzo

Subscribers: vkuzo, supriyar

Tasks: https://github.com/pytorch/pytorch/issues/69535

Differential Revision: [D33405192](https://our.internmc.facebook.com/intern/diff/D33405192)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33405192).
